source/extensions/clusters/original_dst/original_dst_cluster.h

```diff
@@ -110,6 +110,12 @@ class OriginalDstCluster : public ClusterImplBase {
   struct LoadBalancerFactory : public Upstream::LoadBalancerFactory {
     LoadBalancerFactory(const std::shared_ptr<OriginalDstCluster>& cluster) : cluster_(cluster) {}
 
+    ~LoadBalancerFactory() override {
+      // Ensure that the parent cluster and its timer are destroyed on the main thread.
+      std::shared_ptr<OriginalDstCluster> cluster = std::move(cluster_);
```

Comment:
Should we add an ASSERT to the OriginalDstCluster to verify it is deleted on the main thread?

Also is it guaranteed that the cluster_ ref count is 1 at this point? Because if it is not, then posting to main thread may be a no-op and will be deleted at some other point on the wrong thread.

Reply:
Updated with an assertion and a refactor to hopefully capture all references to OriginalDstCluster.